Feature: ui for thesaurus ( issue #61 )

horba opened the pull request:
#61

horba committed:
added thesauruses testing page (view and controller for it);

horba committed:
added translations needed for thesauruses;

horba committed:
added ref to thesauruses testing page into "menu";

horba committed:
added thesaurus directive;

horba committed:
added draft thesaurusService;

horba committed:
Merge remote-tracking branch 'origin/development' into feature/ui-for-thesaurus

horba committed:
path to directive after file relocation fixed;

horba committed:
added missed "return" in remove method;

horba committed:
-added addNewTopic method;

-some watchers deleted ("::" added);

horba committed:
-editing fixed;

-renaming from thesauruse to thesaurus;

-work version without complex structure (as in "cities");

horba committed:
thesaurus directive testing;

horba committed:
changed server url for testing with json-server;

horba committed:
Merge remote-tracking branch 'origin/development' into feature/ui-for-thesaurus

# Conflicts:
#	src/wwwroot/app/main.js
#	src/wwwroot/app/translations/translationsEn.json
#	src/wwwroot/app/translations/translationsRu.json

akozenko committed:
small codestyle

horba committed:
druft complex thesauruses functionality with select list added;

horba committed:
draft editing of selected properties in complex thesauruses added;

horba committed:
changed logic of thesauruses and inner thesauruses loading in thesaurusService (added list to load and run with $q.all);

added  addition fields logic for texts of inner thesauruses;

horba committed:
Merge remote-tracking branch 'origin/feature/ui-for-thesaurus' into feature/ui-for-thesaurus

# Conflicts:
#	src/wwwroot/app/directives/thesaurus/thesaurus.js

horba committed:
changes after merge;

horba committed:
Merge remote-tracking branch 'origin/development' into feature/ui-for-thesaurus

# Conflicts:
#	src/wwwroot/app/index.html
#	src/wwwroot/app/translations/translationsEn.json

horba committed:
fixed bugs after merging;

horba committed:
changes for getting datas from backend (updated structure and server address);

horba committed:
cities ref to country text is name field (not title) - bug fixed;

horba committed:
add filter for showing only not technical info;

horba committed:
filter isShowField added;

horba committed:
added checking for existing of thesaurus name in saving and deleting functions;

horba committed:
bug of getThesaurusNames in thesaurusService fixed (wrong using keys);

horba committed:
small code style changes;

horba committed:
thesauruses view with controller updated (show all thesauruses and only thesauruses);

horba committed:
added select for new topic adding;

horba committed:
relocation THESAURUS_STRUCTURE constant  into single file ThesaurusStructuresStore.js;

akozenko commented on src/wwwroot/app/directives/thesaurus/thesaurus.js:
```diff
@@ -0,0 +1,146 @@
+   function _getThesaurusStructure() {
+      vm.structure = ThesaurusService.getThesaurusStructure(vm.name);
+      forEach(vm.structure.fields,
+               field => {
```
use naming function fot iterator

akozenko commented on src/wwwroot/app/directives/thesaurus/thesaurus.js:
```diff
@@ -0,0 +1,146 @@
+   function _setSelectedObjects(topic) {
+      forEach(_getSelectFields(), field => {
+         vm.selectedObjectsOfEditeTopic[field.name] =
+            find(vm.additionThesaurusesStore[field.refTo], storeTopic =>
```
u can use _.predicate or object ( find(arr, {id: topic[field.name]}) )

akozenko commented on src/wwwroot/app/services/ThesaurusStructuresStore.js:
```diff
@@ -0,0 +1,41 @@
+export const THESAURUS_STRUCTURES = {
```
use default, or use different export for single thesaurus

akozenko commented on src/wwwroot/app/services/thesaurusService.js:
```diff
@@ -0,0 +1,124 @@
+   }
+
+   getThesaurusTopics(thesaurusName) {
+      if (!has(cache, thesaurusName)) {
```
I really dislike `not` in condition.

please, replace it onto

```
if (has(cache, thesaurusName)) {
   return _$q.when(cache[thesaurusName]);
}
....
```

this will be much better

akozenko commented on src/wwwroot/app/services/thesaurusService.js:
```diff
@@ -0,0 +1,124 @@
+               return _entity;
+            });
+      }
+      else {
```
why `else` in new line?

akozenko commented on src/wwwroot/app/services/thesaurusService.js:
```diff
@@ -0,0 +1,124 @@
+function _getLoadedThesaurusesList(mainThesaurusName) {
+   var list = [];
+   list.push(mainThesaurusName);
+   forEach(map(_getReferenceFields(mainThesaurusName), field => field.refTo),
```
it's really hard to read. please, use `let` instead of new line break

akozenko replied:
try this

```
let array = map(_getReferenceFields(mainThesaurusName), 'refTo');
array.unshift(mainThesaurusName);
return array;
```

akozenko commented on src/wwwroot/app/services/thesaurusService.js:
```diff
@@ -0,0 +1,124 @@
+}
+
+function _addRefTextFieldFunction(field, topic, _cache) {
+   var additThesaurus = find(_cache[field.refTo], s => s.id === topic[field.name]);
```
you don't need `find`, you need `some`

akozenko commented on src/wwwroot/app/services/thesaurusService.js:
```diff
@@ -0,0 +1,124 @@
+   }
+}
+
+function _deleteRefTextFieldFunction(field, topic, _cache) {
```
why you don't have any eslint-errors in this line?

akozenko commented on src/wwwroot/app/services/thesaurusService.js:
```diff
@@ -0,0 +1,124 @@
+   return list;
+}
+
+function _addRefTextFieldFunction(field, topic, _cache) {
```
why you use `_cache` as inner var?

akozenko commented on src/wwwroot/app/services/thesaurusService.js:
```diff
@@ -0,0 +1,124 @@
+         let thesaurusesToLoad = _getLoadedThesaurusesList(thesaurusName);
+         return _$q.all(map(thesaurusesToLoad, thesaurus => _HttpService.get(thesaurus)))
+                .then(thesauruses => {
+                   for (var i = 0; i < thesauruses.length; i++) {
```
use `forEach` instead of native `for`

horba replied:
I need to know index not only for thesauruses, but for thesaurusesToLoad too.

akozenko replied:
it's bad practice to use `index` you can use simple object

akozenko commented on src/wwwroot/app/directives/thesaurus/thesaurus.js:
```diff
@@ -0,0 +1,146 @@
+   vm.isEditTopic       = isEditTopic;
+   vm.addNewTopic       = addNewTopic;
+   vm.saveEditTopic     = saveEditTopic;
+   vm.isTopicEditAllow  = isTopicEditAllow;
```
why do you need this function?

horba replied:
what function?

akozenko replied:
`isTopicEditAllow`

horba replied:
For disabling edit and delete buttons if some topic is editing yet

akozenko replied:
ok

akozenko commented on src/wwwroot/app/directives/thesaurus/thesaurus.js:
```diff
@@ -0,0 +1,146 @@
+   }
+
+   function getSelected(id, thesaurusRef) {
+      return find(vm.additionThesaurusesStore[thesaurusRef], s => s.id === id);
```
```
retunr find(vm.additionThesaurusesStore[thesaurusRef], {id: id});
```

akozenko commented on src/wwwroot/app/services/thesaurusService.js:
```diff
@@ -0,0 +1,124 @@
+
+function _getReferenceFields(thesaurusName) {
+   if (has(THESAURUS_STRUCTURES, thesaurusName)) {
+      return filter(THESAURUS_STRUCTURES[thesaurusName].fields, field => has(field, 'refTo'));
```
why not 

```
return filter(THESAURUS_STRUCTURES[thesaurusName].fields, {type: 'select'});
```

?

akozenko replied:
also you use `has` ans `get`. Maybe will better if you use `let`?

horba committed:
code style changes;

some fixes for comments to pull request #90;

akozenko commented on src/wwwroot/app/directives/thesaurus/thesaurus.js:
```diff
@@ -0,0 +1,144 @@
+   }
+
+   function _getSelectFields() {
+      return filter(vm.structure.fields, field => has(field, 'refTo'));
```
why not

```
return filter(vm.structure.fields, {type: 'select'});
```

?

akozenko commented on src/wwwroot/app/directives/thesaurus/thesaurus.js:
```diff
@@ -0,0 +1,144 @@
+   }
+
+   function _onError(message) {
+      vm.message = 'Sorry! Some error occurred: ' + message;
```
use translate

akozenko commented on src/wwwroot/app/services/thesaurusService.js:
```diff
@@ -0,0 +1,123 @@
+      if (has(cache, thesaurusName)) {
+         return _$q.when(cache[thesaurusName]);
+      } else {
+         let thesaurusesToLoad = _getLoadedThesaurusesList(thesaurusName);
```
please, try something like this 

```
let thesaurusesToLoad = _getLoadedThesaurusesList(thesaurusName);
let map = mapValues(thesaurusesToLoad, _HttpService.get);

return _$q.all(map).then(thesauruses => {
   each(thesauruses, (thesaurus, name) => {
      cache[name] = thesaurus;
      _actionOfAdditionFieldsForTopics(thesaurus, name, _addRefTextFieldFunction);
   });
   return cache[thesaurusName];
});

```

akozenko commented on src/wwwroot/app/directives/thesaurus/thesaurus.directive.html:
```diff
@@ -0,0 +1,46 @@
+<div class="errorMessage" ng-bind="message" ng-disabled="!message"></div>
+<table>
+   <caption ng-bind="::structure.thesaurusName | translate"></caption>
+   <tr><th ng-repeat="field in ::structure.fields | filter : isShowField" ng-bind="::field.label"></th></tr>
```
so, you call `field in ::structure.fields | filter : isShowField` three times
it will be good for performance, and for codestyle, if you do this once in your controller

akozenko commented on src/wwwroot/app/directives/thesaurus/thesaurus.directive.html:
```diff
@@ -0,0 +1,46 @@
+   <tr><th ng-repeat="field in ::structure.fields | filter : isShowField" ng-bind="::field.label"></th></tr>
+   <tr ng-repeat = "topic in topics">
+      <td ng-repeat = "field in ::structure.fields | filter : isShowField" >
+         <span ng-if="!isEditTopic(topic)&&field.type==='text'" ng-bind="::topic[field.name]"></span>
```
maybe will be a good idea to use `ng-disabled` instead of render two different element?
and styling disabled element as simple text?

akozenko commented on src/wwwroot/app/services/thesaurusService.js:
```diff
@@ -0,0 +1,123 @@
+         return _$q.when(find(cache[thesaurusName], {id: id}));
+      } else {
+         var url = thesaurusName + '/' + id;
+         return this.getThesaurusTopics(thesaurusName)
```
why you separate this line to 2 lines?

akozenko commented on src/wwwroot/app/services/thesaurusService.js:
```diff
@@ -0,0 +1,123 @@
+      if (has(this.getThesaurusNames(), thesaurusName)) {
+         let additionalUrl = thesaurusName + '/' + entity.id;
+         _actionOfAdditionFieldsForTopic(entity, thesaurusName, _deleteRefTextFieldFunction);
+         return  _HttpService.remove(additionalUrl, entity)
```
use one line 
`return  _HttpService.remove(additionalUrl, entity).then(() => remove(cache[thesaurusName], entity));`

horba replied:
in this case length of line is more then 100 symbols.

msvaillant commented:
@gasty , remember to attach number of issue which this pull-request's features are related

do it in commit or with comments of pr section (not in the header - in this case link will be just a text specification)

horba committed:
change max length of line from 100 to 120 in eslint;

horba committed:
add errors messages into translate;

horba committed:
add style file for thesaurusDirective;

horba committed:
added ng-disabled and deleted spans for showing info in no-edit mode;

horba committed:
code style changes by comments to pull request #90;

horba committed:
Merge remote-tracking branch 'origin/development' into feature/ui-for-thesaurus

horba committed:
added styles for text input and select in disabled mode;

horba committed:
vm.structure replace onto two variables in scope: thesaurusNameLabel (translated name of thesaurus to showing) and fields (filtered for showing fields);

horba committed:
Merge remote-tracking branch 'origin/development' into feature/ui-for-thesaurus

horba committed:
merge development;

changes in structure after changes in backend from development;

fix of incorrect using "has" function to "includes" for checking of thesaurus or not;

akozenko commented on src/wwwroot/app/directives/thesaurus/thesaurus.js:
```diff
@@ -0,0 +1,148 @@
+      ThesaurusService.deleteThesaurusTopic(vm.name, topic).catch(_onError);
+   }
+
+   function _getThesaurusStructure() {
```
this is not `get`

akozenko commented on src/wwwroot/app/directives/thesaurus/thesaurus.js:
```diff
@@ -0,0 +1,148 @@
+      forEach(vm.fields, _fillAdditionThesauruses);
+   }
+
+   function _getThesaurusTopics() {
```
this is not `get`

akozenko commented on src/wwwroot/app/directives/thesaurus/thesaurus.js:
```diff
@@ -0,0 +1,148 @@
+   function _setSelectedObjects(topic) {
+      forEach(_getSelectFields(), field => {
+         vm.selectedObjectsOfEditeTopic[field.name] =
+            find(vm.additionThesaurusesStore[field.refTo], {id: topic[field.name]});
```
why you don't use `getSelected` ?

akozenko commented on src/wwwroot/app/directives/thesaurus/thesaurus.scss:
```diff
@@ -0,0 +1,18 @@
+
+input.edit {
+   &[type="text"] {
+      &:disabled {
```
why not 

```
input.edit[type="text"]:disabled {
```

?

akozenko commented on src/wwwroot/app/services/ThesaurusStructuresStore.js:
```diff
@@ -0,0 +1,43 @@
+         {name : 'id', label : 'id', type : ''},
+         {name : 'title', label : 'name', type : 'text'},
+         {name : 'editTime', label : 'editTime', type : ''},
+         {name : 'state', label : 'state', type : ''}
```
please, use aka-column for better readable

akozenko commented on src/wwwroot/app/services/thesaurusService.js:
```diff
@@ -0,0 +1,138 @@
+      }
+   }
+
+   saveThesaurusTopic(thesaurusName, entity) {
```
try something like this

```

let _curryAction = curry(_actionOfAdditionFieldsForTopic, 3);


................


saveThesaurusTopic(thesaurusName, entity) {
   if (includes(this.getThesaurusNames(), thesaurusName)) {
      let _action = _curryAction(thesaurusName);
      _action(_deleteRefTextFieldFunction, entity);

      let promise;

      if (entity.id) {
         let additionalUrl = thesaurusName + '/' + entity.id;
         promise = _HttpService.put(additionalUrl, entity);
      } else {
         promise = _HttpService.post(thesaurusName + '/', entity);
         promise = promise.then(_entity => {
            cache[thesaurusName].push(_entity);
            return _entity;
         });
      }

      return promise.then(_action(_addRefTextFieldFunction));
   } else {
      return _$q.reject(_$translate.instant('ERRORS.thesaurusErrors.incorrectNameMsg'));
   }
}


......................

_actionOfAdditionFieldsForTopic(thesaurusName, function, entity)



```

horba committed:
issue #61 pull request #90, changes by comments:

scss style fixes;

horba committed:
aka-columns added to ThesaurusStructuresStore;

_getThesaurusTopics renaming into _initThesaurusTopics;

_getThesaurusStructure renaming into _initThesauruseStructure;

getSelected using in _setSelectedObjects added;

horba committed:
refactoring of saveThesaurusTopic();

horba committed:
carring added in _actionOfAdditionFieldsForTopics;

horba committed:
getSelected() calling: wrong arguments list fixed;

akozenko commented on src/wwwroot/app/services/thesaurusService.js:
```diff
@@ -0,0 +1,140 @@
+}
+
+function _addRefTextFieldFunction(field, topic) {
+   if (some(cache[field.refTo], {id: topic[field.name]})) {
```
please, change it for better result

```
   let referencedTopic = find(cache[field.refTo], {id: topic[field.name]});
   if (referencedTopic) {
      topic[field.additionFieldForText] = referencedTopic[field.labelRefFieldName];
   }
```

or

```
let referencedTopic = find(cache[field.refTo], {id: topic[field.name]});
topic[field.additionFieldForText] = (referencedTopic || {})[field.labelRefFieldName];

```

akozenko commented on src/wwwroot/app/services/thesaurusService.js:
```diff
@@ -0,0 +1,140 @@
+}
+
+function _actionOfAdditionFieldsForTopics(topics, thesaurusName, action) {
+   each(topics, _curryAction(thesaurusName, action));
```
please, use something one `each` or `forEach`

https://lodash.com/docs#forEach

horba committed:
Merge remote-tracking branch 'origin/development' into feature/ui-for-thesaurus

# Conflicts:
#	src/wwwroot/.eslintrc
#	src/wwwroot/app/bot-config.js
#	src/wwwroot/app/index.html
#	src/wwwroot/app/main.js

horba committed:
fixed ESLint errors after development merging;

horba committed:
each change into forEach;

horba committed:
changes by comments to pull request #90
